ports.js

```diff
@@ -1,51 +1,228 @@
+        resolve(data[key] || null);
+      });
     });
-});
+  }
+}
 
-app.ports.encryptData.subscribe(function(data) {
-  if (!data.content) {
-    return;
+function setItem(key, value) {
+  if (!IS_WEB_EXTENSION) {
+    if (value === null) {
+      localStorage.removeItem(key);
+    } else {
+      localStorage.setItem(key, value);
+    }
+    return Promise.resolve(null);
+  } else {
+    return new Promise(function(resolve, reject) {
+      let payload = {};
+      payload[key] = value;
+      chrome.storage.local.set(payload, () => {
+        resolve(null);
+      });
+    });
+  }
+}
+
+
+function createElmApp(flags) {
+  let app;
+  if (typeof Elm === 'undefined') {
+    const Elm = require('./hoverpad.js');
+    app = Elm.Main.fullscreen(flags);
+  } else {
+    app = Elm.Main.fullscreen(flags);
   }
-  encrypt(data.passphrase, data.content.replace(/<br>$/g, ''))
-    .then(encryptedContent => {
-      app.ports.dataEncrypted.send(encryptedContent);
-    })
-    .catch(err => {
-      console.error('Error encrypting', err);
-      app.ports.dataNotEncrypted.send(err.message);
+
+  app.ports.getData.subscribe(function() {
+    getItem(CONTENT_KEY)
+      .then(function(data) {
+        app.ports.newData.send([CONTENT_KEY, data || ""]);
+      })
+      .catch(function(err) {
+        console.error(err);
+        app.ports.newError.send('Nothing retrieved: ' + err.message);
+      });
+  });
+
+  app.ports.saveData.subscribe(function(data) {
+    setItem(data.key, data.content)
+      .then(function() {
+        app.ports.dataSaved.send(data.key);
+      })
+      .catch(function(err) {
+        console.error(err);
+        app.ports.newError.send('Nothing retrieved: ' + err.message);
+      });
+  });
+
+  app.ports.savePassphrase.subscribe(function(passphrase) {
+    storePassphrase(passphrase)
+      .then(function() {
+        console.log("Passphrase saved");
+      })
+      .catch(function(err) {
+        console.error(err);
+        app.ports.newError.send('Could not save passphrase: ' + err.message);
+      });
+  });
+
+  app.ports.dropPassphrase.subscribe(function() {
+    dropPassphrase()
+      .then(function() {
+        console.log("Passphrase dropped");
+      })
+      .catch(function(err) {
+        console.error(err);
+        app.ports.newError.send('Could not drop passphrase: ' + err.message);
+      });
+  });
+
+  app.ports.decryptData.subscribe(function(data) {
+    if (!data.content) {
+      app.ports.dataDecrypted.send(null);
+      return;
+    }
+    decrypt(data.passphrase, data.content)
+      .then(content => {
+        app.ports.dataDecrypted.send(content);
+      })
+      .catch(err => {
+        console.error('Error decrypting', err);
+        app.ports.dataNotDecrypted.send(err.message);
+      });
+  });
+
+  app.ports.encryptData.subscribe(function(data) {
+    if (!data.content) {
+      return;
+    }
+    encrypt(data.passphrase, data.content.replace(/<br>$/g, ''))
```

Comment:
Should we go further in sanitizing HTML? there are libs for that, but may be overkill for now.